pulpcore/app/serializers/fields.py

```diff
@@ -19,6 +21,15 @@ def relative_path_validator(relative_path):
         )
 
 
+@extend_schema_field(OpenApiTypes.OBJECT)
+class JSONDictField(serializers.JSONField):
+    """A drf JSONField override to force openapi schema to use 'object' type.
+
+    Not strictly correct, but we relied on that for a long time.
+    See: https://github.com/tfranzel/drf-spectacular/issues/1095
+    """
```

Comment:
My whole point here is: When you _know_ you want dicts there and you go out of your way to specify this particular type explicitly, it is as a matter of fact correct.
When your field "wants" a list and you specify this field, it is just wrong and even the bindings will blow up.
```suggestion
    """A JSONField accepting dicts, specifying as type 'object' in the openapi."""
```

Now writing it this way, i realize, we may even start to add a bit of validation to this class.

Reply:
> When your field "wants" a list and you specify this field, it is just wrong and even the bindings will blow up.

That is the point: it is not "wrong" from the serializer.JSONField (and this subclass) implementation. The serializers will happily parse json booleans, lists and integers, and the database Field will save it on models.JSONField without a problem. Unless we add validation, as you said. Then it would be really wrong, in the serializer level.

Here is my reasoning:
1. We used serializer.JSONField for list and dict-like objects, which are saved correctly in the JSONField db. It is an any json type field.
2. Our bindings worked, because the schema was wrong:
    * It reported `object` type for this Field, when it should report an `any` type (not very accurate for API/bindings purposes).
3. We patch the field's schema representation to preserve the old behavior and not break things right away.
    * This is really supposed to be a patch that we'll remove when we can (on breaking releases, for example).
    * IHO, we shouldnt use this for any new fields, but do (4).
4. Now, on the long run, we fix/replace existing generic JSONType serializer fields.
    * We change only the serializer fields to be strictly dicts/list or whathever (if they are compatible with models.JSONField). This is mainly for more accurate API/schema representation.
    * We may also change the underlying model/db field, if we see a particular reason for it.

Reply:
And to me this is the first instance of a more specific / accurate schema representation.
We would not use it if we saved a list in the json field. But we would need to add other annotations in that case, because (as far as i understood) the current implementation of the bindings cannot consume "Any".

Reply:
Yeah, that could work out too.
Then this PR should probably add that dict validation and add another JSONListField (with list validation and the list schema type), so plugins can use the correct one.

But I'm not 100% sure we can't just use DictField and ListField for dict/list structures instead of creating a new one based on JSONField. I assume from-API desserializing and the to-db serializing should be compatible, even if we keep model.JSONField. I'll test this out just to ensure. If that's the case we dont even need to expose anything at this point.

> because (as far as i understood) the current implementation of the bindings cannot consume "Any".

Yeah, the current Ruby generator version we use can't handle that (newer ones can).

Reply:
Thank you. Adding the validation is probably a good call. If we can just switch to dict field (https://www.django-rest-framework.org/api-guide/fields/#dictfield looks like you can just "not provide the child parameter") that would be even nicer.
Adding a new JSONListField without knowing we are actually going to use it? IDK.

Reply:
>  Adding a new JSONListField without knowing we are actually going to use it? IDK.

Oh, we are using it:
https://github.com/pulp/pulp_rpm/blob/c3c9b1c4582f8e39aab6884951046c55cdf5d5d4/pulp_rpm/app/serializers/package.py#L81-L141

But again, if we can just use serializers.ListField that's better.

Reply:
From the implementation code, DictField's `to_internal_value` and `to_representation` are substantially different from JSONField.
I'll move foward and add JSONDictField and JSONListField with that validation on top and with the spectacular type override, does that sounds ok?